### DISCUSSION TITLE
Telnet to UART gateway got double character on any input

### DISCUSSION THREAD

**Discussion by cwt** [micropython/micropython#15775] [ESP32] [answered]
I connected my ESP32 to the UART port on an SBC because I wanted to connect to its U-Boot.

I ran the following code on my ESP32: https://gist.github.com/cwt/ddf0bb4f4eda86df862c8630629c9857

However, every time I type anything, the characters are displayed twice. For example, `ls` is displayed as `llss`, and this happens instantly before I press enter. BTW, the string (command) sent to the SBC is correct, and the string that the SBC returns via UART and then to my telnet client is also correct without double characters.

What could be the possible problem in my code, and how can I fix it?

**Comment by cwt** [comment]
FYI: It wasn't local echo problem, local echo was disabled on my telnet client.
FYI2: Oh! if I typed fast enough, `ls` is displayed as `lsls` instead of `llss` when typed it slowly.

**Reply by Josverl** [reply]
thans for the report , you can then mark your own answer as The Answer

**Reply by cwt** [reply]
Oh, it wasn't the answer, I just provided more information. I'm still trying to fix the problem

**Comment by shariltumin** [comment, marked as answer]
Please try to add a line (after line 47):

```py
                if conn is None:
                    conn, addr = s.accept()
                    print('Client connected from', addr)
                    _ = conn.recv(256) # ignore connection handshake/header
                    conn.setblocking(False)

```
to ignore the connection "TELNET LINEMODE" negotiation. Maybe this will solve your problem.

**Reply by cwt** [reply]
This help a lot! no more twice characters on typing, but it still echo my input after I hit enter, and also display my password at the password prompt. BTW, this is really much better, and make it usable now.
